assignments/assignment7/faulty-oops.md

```diff
@@ -0,0 +1,75 @@
+Mem abort info:
+  ESR = 0x96000045
+  EC = 0x25: DABT (current EL), IL = 32 bits
+  SET = 0, FnV = 0
+  EA = 0, S1PTW = 0
+  FSC = 0x05: level 1 translation fault
+Data abort info:
+  ISV = 0, ISS = 0x00000045
+  CM = 0, WnR = 1
+user pgtable: 4k pages, 39-bit VAs, pgdp=0000000042639000
+[0000000000000000] pgd=0000000000000000, p4d=0000000000000000, pud=0000000000000000
+Internal error: Oops: 96000045 [#1] SMP
+Modules linked in: hello(O) faulty(O) scull(O)
+CPU: 0 PID: 159 Comm: sh Tainted: G           O      5.15.18 #1
+Hardware name: linux,dummy-virt (DT)
+pstate: 80000005 (Nzcv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
+pc : faulty_write+0x14/0x20 [faulty]
+lr : vfs_write+0xa8/0x2b0
+sp : ffffffc008d23d80
+x29: ffffffc008d23d80 x28: ffffff8002698000 x27: 0000000000000000
+x26: 0000000000000000 x25: 0000000000000000 x24: 0000000000000000
+x23: 0000000040000000 x22: 0000000000000012 x21: 000000558e452a70
+x20: 000000558e452a70 x19: ffffff800266c700 x18: 0000000000000000
+x17: 0000000000000000 x16: 0000000000000000 x15: 0000000000000000
+x14: 0000000000000000 x13: 0000000000000000 x12: 0000000000000000
+x11: 0000000000000000 x10: 0000000000000000 x9 : 0000000000000000
+x8 : 0000000000000000 x7 : 0000000000000000 x6 : 0000000000000000
+x5 : 0000000000000001 x4 : ffffffc0006f7000 x3 : ffffffc008d23df0
+x2 : 0000000000000012 x1 : 0000000000000000 x0 : 0000000000000000
+Call trace:
+ faulty_write+0x14/0x20 [faulty]
+ ksys_write+0x68/0x100
+ __arm64_sys_write+0x20/0x30
+ invoke_syscall+0x54/0x130
+ el0_svc_common.constprop.0+0x44/0xf0
+ do_el0_svc+0x40/0xa0
+ el0_svc+0x20/0x60
+ el0t_64_sync_handler+0xe8/0xf0
+ el0t_64_sync+0x1a0/0x1a4
+Code: d2800001 d2800000 d503233f d50323bf (b900003f)
+---[ end trace dc21db549ccb5a62 ]---
+```
+
+From this output we can see it contains a dump of the kernel panic message due to trying to load the faulty module.
+This contains information about the modules linked, the register contents. If we check the pc (program counter) this tells us
+exactly what instruction was being called in our source code file. In this case the offending line was:
+
+```faulty_write+0x14/0x20 [faulty]```
+
+which is the name of the function which caused the error. We could see it occured 0x14 bytes in to a function that was 0x20 bytes long in a function called faulty_write. Note we can also see this information
+at the top of the call trace which is useful for seeing the called functions leading up to the exception.
+If we check the faulty_write function at https://github.com/cu-ecen-aeld/assignment-7-Silex93/blob/134580cde97cbd7aba9552d39c76fd488e988a95/misc-modules/faulty.c#L49:L55
+we would see:
+
+```
+ssize_t faulty_write (struct file *filp, const char __user *buf, size_t count,
+		loff_t *pos)
+{
+	/* make a simple fault by dereferencing a NULL pointer */
+	*(int *)0 = 0;
+	return 0;
+}
+```
+
+This is the source of the error, we are basically forcing a segfault by accessing an invaild memory location
```

Comment:
@Silex93 when in kernel space it's not typically referred to as a segfault, you'd probably just refer to it as a Kernel Oops.